res/controllers/midi-components-0.1.js

```diff
@@ -0,0 +1,1135 @@
+                // To work around this, reset push Buttons' inKey to 0 when the shift
+                // button is pressed.
+                if (typeof component.shift === "function") {
+                    if (component instanceof Button
+                        && (component.type === Button.types.push
+                            || component.type === undefined)
+                        && component.input === Button.prototype.input
+                        && typeof component.inKey === "string"
+                        && typeof component.group === "string") {
+                        if (engine.getValue(component.group, component.inKey) !== 0) {
+                            engine.setValue(component.group, component.inKey, 0);
+                        }
+                    }
+                    component.shift();
+                }
+            }, false);
+
+            // Shift child ComponentContainers
+            this.forEachComponentContainer(function(container) {
+                container.shift();
+            }, false);
+
+            // Set isShifted for each ComponentContainer recursively
+            this.isShifted = true;
+        }
+        unshift() {
+            // Unshift direct child Components
+            this.forEachComponent(function(component) {
+                // Refer to comment in ComponentContainer.shift() above for explanation
+                if (typeof component.unshift === "function") {
+                    if (component instanceof Button
+                        && (component.type === Button.types.push
+                            || component.type === undefined)
+                        && component.input === Button.prototype.input
+                        && typeof component.inKey === "string"
+                        && typeof component.group === "string") {
+                        if (engine.getValue(component.group, component.inKey) !== 0) {
+                            engine.setValue(component.group, component.inKey, 0);
+                        }
+                    }
+                    component.unshift();
+                }
+            }, false);
+
+            // Unshift child ComponentContainers
+            this.forEachComponentContainer(function(container) {
+                container.unshift();
+            }, false);
+
+            // Unset isShifted for each ComponentContainer recursively
+            this.isShifted = false;
+        }
+        shutdown() {
+            this.forEachComponent(function(component) {
+                if (component.shutdown !== undefined
+                    && typeof component.shutdown === "function") {
+                    component.shutdown();
+                }
+            });
+        }
+    };
+
+    const Deck = class extends ComponentContainer {
+        constructor(deckNumbers) {
+            super();
+            if (deckNumbers !== undefined) {
+                if (Array.isArray(deckNumbers)) {
```

Comment:
this type deduction is a little gross, do we have different scripts doing different things?  Could we fix that?

(I am getting it in my head that we might want to migrate controller to typescript eventually, and the more type dynamism we have the harder that would be)

Reply:
> migrate controller to typescript eventually.

I'm very opposed to this for a number of reasons, but that discussion is out of scope. 

> this type deduction is a little gross, do we have different scripts doing different things? Could we fix that?

In this case, its essentially convenience. You could do the same in most statically typed languages (rust `enum`, haskell `Either`). Instead of having two different parameters, both of which could be unset, set with the correct type or incorrect type as well as being mutually exclusive. We just make use of the fact that those parameters have different types and then switch on type which actually simplifies the code.

Reply:
This is another case of "if you want rigidly defined types, use a different programming language". If you try to bend JS into something rigidly typed, you'll end up with quite strange and unidiomatic JS.